index.d.ts

```diff
@@ -0,0 +1,133 @@
+declare module 'fluture' {
+  interface ResolveFunction<T> {
+    (value: T): void
+  }
+  interface RejectFunction<S> {
+    (error: S): void
+  }
+
+  interface FunctionReturning<T> {
+    (...t: any[]): T
+  }
+
+  interface Cancel {
+    (): void
+  }
+
+  export abstract class IFuture<L, R> {
+    fork (reject: RejectFunction<L>, resolve: ResolveFunction<R>): Cancel
+  }
+
+  export function after<L, R> (duration: number, value: string): IFuture<L, R>
+
+  export function and<L, R> (a: IFuture<L, R>, b: IFuture<L, R>): IFuture<L, R>
+
+
+  // TODO: Need Help
+  export function ap<L, R> (): IFuture<L, R>
+
+  export function attempt<L, R> (f: () => L): IFuture<L, R>
+
+  export function bimap<A, B, AA, BB> (left: (a: A) => AA,
+                                       right: (a: B) => BB): IFuture<AA, any>
+
+  export function both<A, B> (a: IFuture<A, any>, b: IFuture<B, any>): IFuture<[A, B], any>
+
+  export function cache<L, R> (f: IFuture<L, any>): IFuture<L, R>
+
+  export function chain<L, R> (fn: (s: R) => IFuture<IFuture<L, any>, any>, src: IFuture<R, any>): IFuture<L, R>
+
+  export function encase2<L, R, A, B> (fn: (a: A, b: B) => L, a: A, b: B): IFuture<L, R>
+
+  export function encase3<L, R, A, B, C> (fn: (a: A, b: B, c: C) => L, a: A, b: B, c: C): IFuture<L, R>
+
+  export function encase<L, R> (fn: FunctionReturning<L>, ...t: any[]): IFuture<L, R>
+  export function encase<L, R> (fn: FunctionReturning<L>): {(...t: any[]): IFuture<L, R>}
```

Comment:
Again I realized this earlier today :) 
does the following make sense —
```ts
export function encase<L, R> (fn: () => R): {(): IFuture<L, R>}
export function encase<L, R, V> (fn: (v: V) => R, t: V): IFuture<L, R>
export function encase<L, R, V> (fn: (v: V) => R): {(t: V): IFuture<L, R>}
```

Apparently `encase` is polymorphic so we should support all the three cases.